src/Interpreters/GraceHashJoin.cpp

```diff
@@ -410,6 +410,8 @@ void GraceHashJoin::initialize(const Block & sample_block)
 
 void GraceHashJoin::joinBlock(Block & block, std::shared_ptr<ExtraBlock> & not_processed)
 {
+    std::shared_lock current_bucket_lock(current_bucket_mutex);
```

Comment:
If we go with the mutex solution now, then rather need to use `hash_join_mutex` here and remove `current_bucket_mutex` at all since we need to synchronize access to in-memory hash join (plus extra info, like bucket it corresponds to).

Reply:
Probably `current_bucket_mutex` can be removed since there's only one `DelayedJoinedBlocksTransform` instance, will check with another PR